### PR TITLE
DCGEO-171: Add S3 bucket as input to mock server RDA API

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[DCGEO-171](https://jira.cms.gov/browse/DCGEO-171)

**User Story or Bug Summary:**

As a DCGEO developer I need to be able to configure the mock RDA API server to serve data in NDJSON format from an S3 bucket for testing purposes.

---

### What Does This PR Do?

This PR adds a feature to the RDA API mock server to allow it to read claims data from NDJSON files stored in an Amazon S3 bucket.  Accomplishing this led to the following changes:

- Added a new `MessageSource` implementation (`S3JsonMessageSource`) that reads data from a NDJSON file stored in an S3 bucket.
- Added new command line arguments to `RdaServerApp` to configure it to use `S3JsonMessageSources`.
- Moved some of the `S3Utilities` methods into a new class in the bfd-pipeline-shared module (`SharedS3Utilities`) so that it could be re-used in the RDA API pipeline.
- Added command line arguments as an option to `ConfigLoader` and modified `RdaServerApp` to use it for consistency with other apps in the `bfd-pipeline-rda-grpc-server` module.
- Added a `FilteredMessageSource` that allows messages to be filtered to implement replay logic when reading claims from an NDJSON file.
- Made various small changes to files to import and use the classes that were moved during refactoring.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Is this likely to address the goals expressed in the user story?
* Verify all PR security questions and checklists have been completed and addressed.
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* [DCGEO-113 PR 745](https://github.com/CMSgov/beneficiary-fhir-data/pull/745)

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      return true;
+    }
+
+    while (source.hasNext()) {
+      nextValue = source.next();
+      if (filter.test(nextValue)) {
+        return true;
+      }
+    }
+    return false;
```
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java:42]: I was going to comment on sequence number order here, as it could affect the way the filter is being used here, but it's probably reasonable to assume sequence numbers will be in order.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  @Test
+  public void closePassesThrough() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i != 2);
+    filtered.close();
+    verify(source).close();
+  }
+
+  @Test
+  public void multipleHashNextOk() throws Exception {
```
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java:30]: small typo, `multipleHasNextOk()`

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    verify(source, times(2)).hasNext();
+  }
+
+  @Test(expected = NoSuchElementException.class)
+  public void nextWithoutHasNextFails() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i < 10);
+    filtered.next();
```
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java:48]: Should this happen?  I think the normal behavior of iterator type APIs is that `next()` will always handle advancing the pointer as well as grabbing the next value, while `hasNext()` simply checks if the cursor is currently pointing to `null`, acting as a simple safeguard with no side effects.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * Wrap the specified source and filter its objects using the supplied predicate. The predicate
+   * must return true for objects that are valid and false for objects that are invalid.
+   *
+   * @param source real source of objects
+   * @param filter function returning true for objects to keep or false for objects to skip
+   */
+  public FilteredMessageSource(MessageSource<T> source, Predicate<T> filter) {
+    this.source = source;
+    this.filter = filter;
```
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java:26]: Should we advance the pointer here so that `next()` can be called?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    assertEquals(Integer.valueOf(1), filtered.next());
+
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    verify(source, times(2)).hasNext();
+  }
+
+  @Test(expected = NoSuchElementException.class)
+  public void nextWithoutHasNextFails() throws Exception {
```
Comment (KyRobbins) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java:45]: Should we add a test to check if an exception is thrown when we are at the end of the iterator?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java:42]: I agree with your logic and the thought that led to the question. This class is independent of sequence numbers so the order doesn't come into play here.  Generally the message sources needs to be provided with data in proper order.  For example JsonMessageSource doesn't sort the file because that could be very slow.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java (hunk 3 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java:48]: The protocol here is that `hasNext()` has to be called before `next()`.  This isn't an `Iterator` so I feel free to deviate from that interface's protocol.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java (hunk 4 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java:26]: No it's part of the design that the constructor doesn't do any IO.  That happens in `hasNext()`.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/extract/s3/S3Utilities.java [hunk 6]
```diff
@@ -1,70 +1,20 @@
...
-
-    return AmazonS3ClientBuilder.standard()
-        .withEndpointConfiguration(
-            new AwsClientBuilder.EndpointConfiguration(
-                minioConfig.minioEndpointAddress, awsS3Region.name()))
-        .withPathStyleAccessEnabled(true)
-        .withClientConfiguration(clientConfiguration)
-        .withCredentials(new AWSStaticCredentialsProvider(credentials))
-        .build();
+    return SharedS3Utilities.createS3Client(options.getS3Region());
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/extract/s3/S3Utilities.java:18]: I retained this method because it accepts a class local to this module as an argument and it's used in 10 different places.  It's nice not to expand out the options.getS3Region() call in all of those places since it was previously an encapsulated detail.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/RdaServerApp.java [hunk 7]
```diff
@@ -46,35 +50,27 @@ public static void main(String[] args) throws Exception {
...
-          throw new IOException("invalid argument: " + arg);
-        }
+      final ConfigLoader config =
+          ConfigLoader.builder().addKeyValueCommandLineArguments(args).build();
+      port = config.intValue("port", 5003);
+      seed = config.longOption("seed").orElseGet(System::currentTimeMillis);
+      maxToSend = config.intValue("maxToSend", 5_000);
+      fissClaimFile = config.readableFileOption("fissFile").orElse(null);
+      mcsClaimFile = config.readableFileOption("mcsFile").orElse(null);
+      final Optional<String> s3Bucket = config.stringOption("s3Bucket");
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/RdaServerApp.java:63]: thoughts about using a path style option here rather than just the bucket name.  I'm thinking that with the work the BFD team did last sprint in adding minio support, we could test this out locally using a minio path running an a local docker network?

Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/RdaServerApp.java:63]: Yes I have tested this using Minio locally.  A bucket maps to a directory inside the minio data directory.  It doesn't allow path characters in the bucket name.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java (hunk 5 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java:45]: Thanks. I have added that test.

**Review by ericbuckley — APPROVED** [review]
🎉

**Review by KyRobbins — APPROVED** [review]

**Review by christrotteradhoc — APPROVED** [review]
